mitmproxy/command.py

```diff
@@ -159,23 +163,29 @@ def parse_partial(
         """
             Parse a possibly partial command. Return a sequence of ParseResults and a sequence of remainder type help items.
         """
-        buf = io.StringIO(cmdstr)
         parts: typing.List[str] = []
-        lex = lexer(buf)
-        while 1:
-            remainder = cmdstr[buf.tell():]
-            try:
-                t = lex.get_token()
-            except ValueError:
-                parts.append(remainder)
-                break
-            if not t:
-                break
-            parts.append(t)
+
+        rex = pyparsing.QuotedString("\"", escChar='\\', unquoteResults=False) |\
+            pyparsing.QuotedString("'", escChar='\\', unquoteResults=False) |\
+            pyparsing.Combine(pyparsing.Literal('"') + pyparsing.Word(pyparsing.printables + " ") + pyparsing.StringEnd()) |\
+            pyparsing.Word(pyparsing.printables) |\
+            pyparsing.Word(' ')
+
+        rex = rex.copy().leaveWhitespace()
+
+        for t, start, end in rex.scanString(cmdstr):
+            parts.append(t[0])
```

Comment:
Perf 4: If we only use the element itself, we can just go with `parseString`.

Reply:
So, three options there. If I changes this back to parseString I either have to handle an exception for the case that the string is empty or I have to check that the string is not empty prior to calling parseString. The third option is leaving it as a loop. Any preferences?

Reply:
How about we just do a simple `if not cmdstr or cmdstr.isspace(): return []` at the start? :)

Reply:
That function ended up becoming useless at the end after moving the regex object out of it. It ended up being:

```python
        parts: typing.List[str] = []
        for t, start, end in self.regex.scanString(cmdstr):
            parts.append(t[0])
        return parts
```

So I put that loop back into `parse_partial` and cached it instead (which should give a better gain, considering it does more stuff)